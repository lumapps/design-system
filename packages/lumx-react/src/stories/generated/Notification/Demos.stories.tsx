/**
 * File generated when storybook is started. Do not edit directly!
 */
export default { title: 'LumX components/notification/Demos' };

export { App as Default } from './default';
