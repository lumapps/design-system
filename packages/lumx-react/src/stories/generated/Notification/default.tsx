/home/gcornut/git/design-system/packages/site-demo/content/product/components/notification/react/default.tsx