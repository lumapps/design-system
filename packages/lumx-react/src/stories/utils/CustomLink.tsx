/* eslint-disable jsx-a11y/anchor-has-content,jsx-a11y/anchor-is-valid */
import React from 'react';

/**
 * Example custom link that can be used in linkAs props.
 */
export const CustomLink: any = ({ ...props }) => <a href="#" data-custom-link {...props} />;
