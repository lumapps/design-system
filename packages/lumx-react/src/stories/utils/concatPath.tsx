export type PathPart = string | number;
export type Path = PathPart | Array<PathPart>;

/**
 * Concat flatten object path
 *
 * @example concatPath('foo', 'bar') // => 'foo.bar'
 * @example concatPath(['foo', 0]) // => 'foo[0]'
 * @example concatPath('foo', 0, ['bar']) // => 'foo[0].bar'
 */
export const concatPath = (...prefix: Path[]) => {
    const [first, ...rest] = prefix.flat();
    return rest.reduce<string>((acc, part) => {
        if (typeof part === 'number') return `${acc}[${part}]`;
        return `${acc}.${part}`;
    }, String(first));
};
