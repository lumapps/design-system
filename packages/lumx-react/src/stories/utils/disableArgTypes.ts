export function disableArgTypes(argTypes: string[]) {
    return Object.fromEntries(argTypes.map((argType) => [argType, { table: { disable: true } }]));
}
