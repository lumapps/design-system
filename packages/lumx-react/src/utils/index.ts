/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
