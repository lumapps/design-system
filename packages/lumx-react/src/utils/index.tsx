export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
