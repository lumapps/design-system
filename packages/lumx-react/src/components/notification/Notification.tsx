import React, { forwardRef } from 'react';
import { createPortal } from 'react-dom';

import classNames from 'classnames';

import isFunction from 'lodash/isFunction';

import { Button, Emphasis, Icon, Kind, Size, Theme } from '@lumx/react';

import { DOCUMENT, NOTIFICATION_TRANSITION_DURATION } from '@lumx/react/constants';
import { NOTIFICATION_CONFIGURATION } from '@lumx/react/components/notification/constants';
import { Comp, GenericProps, getRootClassName, handleBasicClasses } from '@lumx/react/utils';

import { useDelayedVisibility } from '@lumx/react/hooks/useDelayedVisibility';

/**
 * Defines the props of the component.
 */
export interface NotificationProps extends GenericProps {
    /** Action button label. */
    actionLabel?: string;
    /** Content. */
    content?: React.ReactNode;
    /** Whether the component is open or not. */
    isOpen?: boolean;
    /** Theme adapting the component to light or dark background. */
    theme?: Theme;
    /** Notification type. */
    type?: Kind;
    /** Z-axis position. */
    zIndex?: number;
    /** On action button click callback. */
    onActionClick?(): void;
    /** On click callback. */
    onClick?(): void;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'Notification';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<NotificationProps> = {
    theme: Theme.light,
    zIndex: 9999,
};

/* eslint-disable react-hooks/rules-of-hooks, jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */
/**
 * Notification component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const Notification: Comp<NotificationProps, HTMLDivElement> = forwardRef((props, ref) => {
    const {
        actionLabel,
        className,
        content,
        isOpen,
        onActionClick,
        onClick,
        theme,
        type,
        zIndex,
        ...forwardedProps
    } = props;
    if (!DOCUMENT) {
        // Can't render in SSR.
        return null;
    }
    const { color, icon } = NOTIFICATION_CONFIGURATION[type as Kind] || {};
    const isVisible = useDelayedVisibility(!!isOpen, NOTIFICATION_TRANSITION_DURATION);
    const hasAction: boolean = Boolean(onActionClick) && Boolean(actionLabel);

    const handleCallback = (evt: React.MouseEvent) => {
        if (isFunction(onActionClick)) {
            onActionClick();
        }
        evt.stopPropagation();
    };

    return type && isVisible
        ? createPortal(
              <div
                  ref={ref}
                  role="alert"
                  {...forwardedProps}
                  className={classNames(
                      className,
                      handleBasicClasses({
                          color,
                          hasAction,
                          isHidden: !isOpen,
                          prefix: CLASSNAME,
                      }),
                  )}
                  onClick={onClick}
                  style={{ zIndex }}
              >
                  <div className={`${CLASSNAME}__icon`}>
                      <Icon icon={icon} size={Size.s} />
                  </div>
                  <div className={`${CLASSNAME}__content`}>{content}</div>
                  {hasAction && (
                      <div className={`${CLASSNAME}__action`}>
                          <Button emphasis={Emphasis.medium} theme={theme} onClick={handleCallback}>
                              <span>{actionLabel}</span>
                          </Button>
                      </div>
                  )}
              </div>,
              document.body,
          )
        : null;
});
Notification.displayName = COMPONENT_NAME;
Notification.className = CLASSNAME;
Notification.defaultProps = DEFAULT_PROPS;
