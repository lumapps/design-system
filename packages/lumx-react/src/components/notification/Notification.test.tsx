import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';

import noop from 'lodash/noop';
import { Kind } from '@lumx/react';
import { Wrapper } from '@lumx/react/testing/utils';

import { Notification, NotificationProps } from './Notification';

const CLASSNAME = Notification.className as string;

type SetupProps = Partial<NotificationProps>;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = (propsOverride: SetupProps = {}, shallowRendering = true) => {
    const props: any = { ...propsOverride };
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;
    const wrapper: Wrapper = renderer(<Notification {...props} />);

    const notification: Wrapper = wrapper.find('.lumx-notification');
    const icon: Wrapper = wrapper.find('.lumx-notification__icon');
    const content: Wrapper = wrapper.find('.lumx-notification__content');
    const action: Wrapper = wrapper.find('.lumx-notification__action');

    return { action, content, icon, notification, props, wrapper };
};

const properties = {
    error: {
        content: 'Error',
        onClick: noop,
        isOpen: true,
        type: Kind.error,
    },
    info: {
        content: 'Info',
        onClick: noop,
        isOpen: true,
        type: Kind.info,
    },
    infoWithCallback: {
        onActionClick: noop,
        actionLabel: 'Undo',
        content: 'Info with callback',
        onClick: noop,
        isOpen: true,
        type: Kind.info,
    },
    success: {
        content: 'Success',
        onClick: noop,
        isOpen: true,
        type: Kind.success,
    },

    warning: {
        content: 'Warning',
        onClick: noop,
        isOpen: true,
        type: Kind.warning,
    },
};

describe(`<${Notification.displayName}>`, () => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        it('should render defaults', () => {
            const { wrapper, notification } = setup(properties.info);
            expect(wrapper).toMatchSnapshot();

            expect(notification).toExist();

            expect(notification).toHaveClassName(CLASSNAME);
            expect(notification).toHaveClassName(`${CLASSNAME}--color-dark`);
        });

        it('should render nothing since the notification is closed', () => {
            const { wrapper, notification } = setup({ ...properties.info, isOpen: false });
            expect(wrapper).toMatchSnapshot();

            expect(notification).not.toExist();
        });
    });

    // 2. Test defaultProps value and important props custom values.
    describe('Props', () => {
        // Nothing to do here.
    });

    // 3. Test events.
    describe('Events', () => {
        // Nothing to do here.
    });

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', () => {
        // Nothing to do here.
    });

    // 5. Test state.
    describe('State', () => {
        // Nothing to do here.
    });
});
