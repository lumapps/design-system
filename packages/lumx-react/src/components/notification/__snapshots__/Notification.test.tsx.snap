// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notification> Snapshots and structure should render defaults 1`] = `
<Portal
  containerInfo={<body />}
>
  <div
    className="lumx-notification lumx-notification--color-dark"
    onClick={[Function]}
    role="alert"
    style={
      Object {
        "zIndex": 9999,
      }
    }
  >
    <div
      className="lumx-notification__icon"
    >
      <Icon
        icon="M13,9H11V7H13M13,17H11V11H13M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2Z"
        size="s"
      />
    </div>
    <div
      className="lumx-notification__content"
    >
      Info
    </div>
  </div>
</Portal>
`;

exports[`<Notification> Snapshots and structure should render nothing since the notification is closed 1`] = `""`;
